ui/src/ducks/app/monitoring.js

```diff
@@ -725,6 +743,33 @@ export function* fetchNodeStats({ payload }) {
   yield put(updateNodeStatsAction({ metrics: metrics }));
 }
 
+export function* refreshNodeStats() {
+  const instanceIP = yield select(instanceIPSelector);
+  const controlPlaneInterface = yield select(controlPlaneInterfaceSelector);
+  const workloadPlaneInterface = yield select(workloadPlaneInterfaceSelector);
+
+  yield put(updateNodeStatsAction({ isRefreshing: true }));
+  if (instanceIP && controlPlaneInterface && workloadPlaneInterface) {
+    yield call(fetchNodeStats, {
+      payload: {
+        instanceIP: instanceIP,
+        controlPlaneInterface: controlPlaneInterface,
+        workloadPlaneInterface: workloadPlaneInterface,
+      },
+    });
+  }
+  yield delay(REFRESH_METRCIS_GRAPH);
+
+  const isRefreshing = yield select(isNodeStatsRefreshing);
+  if (isRefreshing) {
+    yield call(refreshNodeStats);
+  }
+}
```

Comment:
This pattern for refresh sagas was already discussed with Carlito, back in the days, and we came to the conclusion that it had big drawbacks (you can loose a "STOP" action, or trigger the refresh loop multiple times in parallel, and overall, it's not easy to read). See #2035.

In practice, with the suggestion from a previous comment, that would look like:

```javascript
export function* refreshNodeStats() {
  while (true) {
    yield take(REFRESH_NODESTATS);
    while (true) {
      yield fork(fetchNodeStats);
      const {interrupt, update} = yield race({
        interrupt: take(STOP_REFRESH_NODESTATS),
        // If the refresh period expires before we receive a halt,
        // we can refresh the stats
        requeue: delay(DELAY_TIMEOUT),
        // NOTE: This action doesn't exist, but it should, so that
        // whenever we change one of the parameters for "fetchNodeStats",
        // it gets triggered again
        update: take(UPDATE_SELECTED_NODE),
      });
      if (interrupt) {
        yield cancel(fetchNodeStats);
        break;
      }
    }
  }
}
```

Reply:
hello @gdemonet ,

I wanted to address this dept ticket for a long time! Let's work on it now!

I have one questions in my mind:
What does `yield take(REFRESH_NODESTATS);` do? In this case, I think we'll never have REFRESH_NODESTATS dispatched.

Reply:
`take(SOME_EVENT)` is blocking until `SOME_EVENT` is seen.

So, the idea is to replace the `takeFirst` effect creator:

```javascript
while (true) {
  // wait for the start event
  yield take("START_REFRESH");
  // once we've seen this start event, we can enter the refresh loop
  while (true) {
    // trigger the data update
    yield fork(fetchData);
    // wait for the stop event for at most the refresh period
    const {interrupt} = yield race({
      interrupt: take("STOP_REFRESH"),
      requeue: delay(REFRESH_PERIOD),
    });
    if (interrupt) {
      // we received the stop event, so we get out of the refresh loop
      // and we will wait for the next start event
      break;
    }
  }
}
```